crates/red_knot_python_semantic/src/types.rs

```diff
@@ -245,46 +263,77 @@ fn definition_expression_ty<'db>(
-        |BindingWithConstraints {
-             binding,
-             constraints,
-         }| {
-            let mut constraint_tys = constraints
-                .filter_map(|constraint| narrowing_constraint(db, constraint, binding))
-                .peekable();
-
-            let binding_ty = binding_ty(db, binding);
-            if constraint_tys.peek().is_some() {
-                constraint_tys
-                    .fold(
-                        IntersectionBuilder::new(db).add_positive(binding_ty),
-                        IntersectionBuilder::add_positive,
-                    )
-                    .build()
-            } else {
-                binding_ty
+) -> Symbol<'db> {
+    let mut is_unbound_visible = false;
+    let mut types = vec![];
+
+    for BindingWithConstraints {
+        binding,
+        constraints,
+        all_constraints,
+        visibility_constraints,
+        visibility_constraint,
+    } in bindings_with_constraints
+    {
+        let static_visibility =
+            visibility_constraints.analyze(db, all_constraints, visibility_constraint);
+
+        let Some(binding) = binding else {
+            if !static_visibility.is_always_false() {
+                // TODO: also distinguish between always-true and ambiguous?
+                is_unbound_visible = true;
             }
-        },
-    );
+            continue;
+        };
 
-    if let Some(first) = def_types.next() {
-        if let Some(second) = def_types.next() {
-            Some(UnionType::from_elements(
-                db,
-                [first, second].into_iter().chain(def_types),
-            ))
+        if static_visibility.is_always_false() {
+            continue;
+        }
+
+        let mut constraint_tys = constraints
+            .filter_map(|constraint| narrowing_constraint(db, constraint, binding))
+            .peekable();
+
+        let binding_ty = binding_ty(db, binding);
+        let ty = if constraint_tys.peek().is_some() {
+            let intersection_ty = constraint_tys
+                .fold(
+                    IntersectionBuilder::new(db).add_positive(binding_ty),
+                    IntersectionBuilder::add_positive,
+                )
+                .build();
+            intersection_ty
+        } else {
+            binding_ty
+        };
+
+        types.push(ty);
+    }
+
+    let mut types = types.into_iter();
+
+    if let Some(first) = types.next() {
+        let boundness = if is_unbound_visible {
+            Boundness::PossiblyUnbound
         } else {
-            Some(first)
+            Boundness::Bound
+        };
+
+        if let Some(second) = types.next() {
+            Symbol::Type(
+                UnionType::from_elements(db, [first, second].into_iter().chain(types)),
+                boundness,
+            )
+        } else {
+            Symbol::Type(first, boundness)
         }
     } else {
-        None
+        Symbol::Unbound
```

Comment:
Maybe a debug assertion here that `is_unbound_visible`?

(If we tracked always-true visible vs ambiguous-visible, I think it should be always-true visible here?)

Reply:
We discussed this separately. The short summary is that we can't add this assertions, as there can be situations in which we call `declarations_ty` (or `bindings_ty`?) in type inference of unreachable code, where this assertion is not fulfilled.


The longer explanation is this: in an example like
```py
(x := 0) and (x := 1) and (x := "foo")
```
- when we infer types for that named expression `x := "foo"`, we call `declarations_ty` to make sure that we can assign `"foo"` to the declared type of `x`.
- the only possible control flow path that leads to `x := "foo"` goes through the other two expressions. 
- along this path, we apply visibility constraints. we apply a visibility constraint of `(x := 0)` to the `(x := 1)` binding, because it is only visible if that first condition is not always-false (which it is here) 
- The `(x := 1)` expression is unreachable code
- When we apply that `(x := 0)` visibility constraint, we also apply it to the scope-start binding/declaration.
- so now if we ask for the declared type when inferring types for `(x := "foo")`, we see the scope-start with a visibility constraint of `(x := 0)`, i.e. always-false
- the scope start is not "reachable" from there, but that is fine